src/Symfony/Component/Workflow/MarkingStore/AbstractMarkingStore.php

```diff
@@ -0,0 +1,155 @@
+        if (!in_array($strategy, $strategies, true)) {
+            throw new InvalidArgumentException(sprintf('Marking strategy must be one of "%s", but got "%".', implode('" or "', $strategies), $strategy));
+        }
+
+        $this->strategy = $strategy;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    final public function getStrategy()
+    {
+        return $this->strategy;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws InvalidMarkingException When the marking does not match the strategy
+     * @throws InvalidMarkingStrategyException When the strategy in unknown
+     */
+    public function getMarking($subject)
+    {
+        if (MarkingInterface::STRATEGY_MULTIPLE_STATE === $this->strategy) {
+            $marking = $this->getMultipleStateMarking($subject);
+
+            if (!$marking instanceof MultipleStateMarking) {
+                throw new InvalidMarkingException(MultipleStateMarking::class, $marking);
+            }
+
+            return $marking;
+        }
+
+        if (MarkingInterface::STRATEGY_SINGLE_STATE === $this->strategy) {
+            $marking = $this->getSingleStateMarking($subject);
+
+            if (!$marking instanceof SingleStateMarking) {
+                throw new InvalidMarkingException(SingleStateMarking::class, $marking);
+            }
+
+            return $marking;
+        }
+
+        throw new InvalidMarkingStrategyException(static::class);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws InvalidMarkingException When the marking does not match the strategy
+     * @throws InvalidMarkingStrategyException When the strategy in unknown
+     */
+    final public function setMarking($subject, MarkingInterface $marking)
+    {
+        if (MarkingInterface::STRATEGY_MULTIPLE_STATE === $this->strategy) {
+            if (!$marking instanceof MultipleStateMarking) {
+                throw new InvalidMarkingException(MultipleStateMarking::class, $marking);
+            }
+
+            $this->setMultipleStateMarking($subject, $marking);
+
+            return;
+        }
+
+        if (MarkingInterface::STRATEGY_SINGLE_STATE === $this->strategy) {
+            if (!$marking instanceof SingleStateMarking) {
+                throw new InvalidMarkingException(SingleStateMarking::class, $marking);
+            }
+
+            $this->setSingleStateMarking($subject, $marking);
+
+            return;
+        }
+
+        throw new InvalidMarkingStrategyException(static::class);
+    }
+
+    /**
+     * Gets a SingleStateMarking from a subject.
+     *
+     * @param object $subject A subject
+     *
+     * @return SingleStateMarking The marking
+     */
+    abstract protected function getSingleStateMarking($subject);
+
+    /**
+     * Sets a SingleStateMarking to a subject.
+     *
+     * @param object             $subject A subject
+     * @param SingleStateMarking $marking
+     *
+     * @return SingleStateMarking The marking
```

Comment:
setters have no reason to return the marking (the return value is ignored anyway)